src/common/utxobased/engine/makeUtxoEngineState.ts

```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
         addressIndex: getAddressCount()
       }
       const { address } = walletTools.getAddress(path)
-      const scriptPubkey = walletTools.addressToScriptPubkey(address)
       await saveAddress({
         ...args,
-        scriptPubkey,
+        address,
         path
       })
 
-      // TODO: don't process addresses during setLookAhead. Addresses should be added to a queue here
-      await processAddress({ ...args, address })
+      taskCache.addressSubscribeCache.set(address, { format, branch })
     }
   })
 }
 
+const addToTransactionCache = async (
+  args: CommonArgs,
+  address: string,
+  format: CurrencyFormat,
+  branch: number,
+  transactions: Map<string, AddressTransactionCacheState>
+): Promise<void> => {
+  const { walletTools, processor } = args
+  // Fetch the networkQueryVal from the database
+  const scriptPubkey = walletTools.addressToScriptPubkey(address)
+  const { networkQueryVal = 0 } =
+    (await processor.fetchAddressByScriptPubkey(scriptPubkey)) ?? {}
+  transactions.set(address, {
+    fetching: false,
+    path: {
+      format,
+      branch
+    },
+    page: 1, // Page starts on 1
+    networkQueryVal
+  })
+}
+
+interface NextTaskArgs extends CommonArgs {
+  blockBook: BlockBook
+}
+
+export const pickNextTask = async (
+  args: NextTaskArgs
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+): Promise<WsTask<any> | undefined> => {
+  const { addressesToWatch, taskCache, blockBook } = args
+
+  const {
+    addressSubscribeCache,
+    utxosCache,
+    rawUtxosCache,
+    processedUtxosCache,
+    transactionsCache,
+    updateTransactionsCache
+  } = taskCache
+
+  console.log({
+    addressSubscribeCache: addressSubscribeCache.size,
+    utxosCache: utxosCache.size,
+    rawUtxosCache: rawUtxosCache.size,
+    processedUtxosCache: processedUtxosCache.size,
+    transactionsCache: transactionsCache.size,
+    updateTransactionsCache: updateTransactionsCache.size
+  })
+
+  // first check if blocks are already being watched
+  if (!taskCache.blockWatching) {
+    taskCache.blockWatching = true
+    const deferred = new Deferred<unknown>()
+    deferred.promise.catch(() => {
+      taskCache.blockWatching = false
+    })
+
+    blockBook.watchBlocks(
+      async () => await onNewBlock({ ...args, blockBook }),
+      deferred
+    )
+  }
+
+  // Loop to process addresses to utxos
+  for (const [address, state] of utxosCache) {
+    // Check if we need to fetch address UTXOs
+    if (!state.fetching) {
+      state.fetching = true
+
+      utxosCache.delete(address)
+
+      // Fetch and process address UTXOs
+      const wsTask = await processAddressUtxos({
+        ...args,
+        ...state,
+        address
+      })
+      return wsTask
+    }
+  }
+
+  // Check if there are any addresses pending to be subscribed
```

Comment:
Subscribe to none-used addresses first, then fetch the transactions, then to used addresses.

Reply:
This requires some change in setLookAhead by the looks of it to distinguish between used and non-used transactions. Subscribe is anyway only on an array, so not sure what benefit this brings.